### PR TITLE
[JUJU-2550] ChangeLog DDL

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following adds the change log DDL schema. The changelog represents a way to collate the flow of changes from a set of table triggers to a stream of data changes. These data changes represented in the change log can be queried at a later stage to know what's changed. This information will be utilised to drive an event stream for watchers to know what's changed.

This will be the sub-system for driving the new juju watchers. This schema was derived from the nu-juju-watchers[1] prototype.

The difference between this change and the prototype, is that we're constraining values to give better type information and indexing opportunities.

Additionally, I've fixed --build-agent when bootstrapping and prevent github actions from running when a new commit is present. You can always manually run the github action if you want the prior one to succeed.

  1. https://github.com/SimonRichardson/nu-juju-watchers/blob/main/schema.go

-----

NOTE: I've switched from a `uuid` as a primary key to an auto-incrementing primary key. This is so we can use sqlite as the official arbiter of ordering for changes occurring. The downside is the following (taken from [2])

   > The AUTOINCREMENT keyword imposes extra CPU, memory, disk space, and disk I/O overhead and should be avoided if not strictly needed. It is usually not needed.

And special mention to

 > If the AUTOINCREMENT keyword appears after INTEGER PRIMARY KEY, that changes the automatic ROWID assignment algorithm to prevent the reuse of ROWIDs over the lifetime of the database. In other words, the purpose of AUTOINCREMENT is to prevent the reuse of ROWIDs from previously deleted rows.

 2. https://www.sqlite.org/autoinc.html

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc

## QA steps

```sh
$ make juju
$ juju bootstrap lxd test --build-agent
```

### CODE REVIEW

File: database/schema/controller.go [hunk 1]
```diff
@@ -56,6 +68,42 @@ ON lease_pin (lease_uuid, entity_id);
...
+CREATE UNIQUE INDEX idx_change_log_namespace_name
+ON change_log_namespace (name);
 
-	return strings.Split(delta, ";\n\n")
+CREATE TABLE change_log (
+    uuid                    TEXT PRIMARY KEY,
+    change_log_type_id      INT NOT NULL,
+    change_log_namespace_id INT NOT NULL,
+    namespace_id            TEXT NOT NULL,
+    created_at              DATETIME NOT NULL,
```
Comment (manadart) [inline database/schema/controller.go:101]: Might be worth adding `DEFAULT CURRENT_TIMESTAMP`, to save on setting explicitly for every trigger...

**Review by manadart — COMMENTED** [review]
Some suggestions.

File: database/schema/controller.go [hunk 2]
```diff
@@ -56,6 +68,42 @@ ON lease_pin (lease_uuid, entity_id);
...
+    type TEXT
+);
+
+CREATE UNIQUE INDEX idx_change_log_type_type
+ON change_log_type (type);
+
+INSERT INTO change_log_type VALUES
+    (1, 'create'),
+    (2, 'update'),
+    (4, 'delete');
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply database/schema/controller.go:86]: Bitwise

File: database/schema/controller.go (hunk 1 above)
Comment (SimonRichardson) [reply database/schema/controller.go:101]: Yeah

Comment (SimonRichardson) [reply database/schema/controller.go]: Actually, I've changed this to use milliseconds now. This should give us better fidelity when grouping in the change stream.

File: database/schema/controller.go [hunk 3]
```diff
@@ -56,6 +68,48 @@ ON lease_pin (lease_uuid, entity_id);
...
+
+CREATE UNIQUE INDEX idx_change_log_namespace_namespace
+ON change_log_namespace (namespace);
+
+-- TODO (stickupkid): Add the namespaces we want to track.
+-- INSERT INTO change_log_namespace VALUES
+--    (1, 'foo');
 
-	return strings.Split(delta, ";\n\n")
+CREATE TABLE change_log (
```
Comment (manadart) [inline database/schema/controller.go:102]: After we contrive the base watcher, we need an exercise in creating the correct indexes based on the access patterns.

**Review by manadart — APPROVED** [review]

File: database/schema/controller.go (hunk 1 above)
Comment (manadart) [reply database/schema/controller.go]: I think it's actually microseconds (2023-02-09 08:32:08.053).

**Pull request merged**